flocker/control/_model.py

```diff
@@ -17,27 +17,71 @@
 from characteristic import attributes
 
 from twisted.python.filepath import FilePath
+
 from pyrsistent import (
     pmap, PRecord, field, PMap, CheckedPSet, CheckedPMap,
+    optional as optional_type
     )
 
 from zope.interface import Interface, implementer
 
 
-def pset_field(klass):
+def pset_field(item_type, optional=False):
     """
-    Create checked ``PSet`` field that can serialize recursively.
+    Create checked ``PSet`` field.
+
+    :param item_type: The required type for the items in the set.
+    :param bool optional: If true, ``None`` can be used as a value for
+        this field.
 
     :return: A ``field`` containing a ``CheckedPSet`` of the given type.
     """
     class TheSet(CheckedPSet):
-        __type__ = klass
-    TheSet.__name__ = klass.__name__ + "PSet"
-
-    return field(type=TheSet, factory=TheSet.create, mandatory=True,
+        __type__ = item_type
+    TheSet.__name__ = item_type.__name__ + "PSet"
+
+    if optional:
+        def factory(argument):
+            if argument is None:
+                return None
+            else:
+                return TheSet(argument)
+    else:
+        factory = TheSet
+    return field(type=optional_type(TheSet) if optional else TheSet,
+                 factory=factory, mandatory=True,
                  initial=TheSet())
 
 
+def pmap_field(key_type, value_type, optional=False):
+    """
+    Create a checked ``PMap`` field.
+
+    :param key: The required type for the keys of the map.
+    :param value: The required type for the values of the map.
+    :param bool optional: If true, ``None`` can be used as a value for
+        this field.
+
+    :return: A ``field`` containing a ``CheckedPMap``.
+    """
+    class TheMap(CheckedPMap):
+        __key_type__ = key_type
+        __value_type__ = value_type
+    TheMap.__name__ = key_type.__name__ + value_type.__name__ + "PMap"
```

Comment:
I'm tempted to say something about underscores.  What's worse, `unicodeFilePathPMap` or `unicode_FilePath_PMap`?  Hard to say.  :question: